File: apps/dashboard/types/incident.ts
export type IncidentStatus =
  | 'created'
  | 'received'
  | 'assigned'
  | 'responding'
  | 'arrived'
  | 'resolved'
  | 'cancelled'
  | 'escalated'
  | 'unassigned'
  | 'failed';

export type EmergencyType =
  | 'medical'
  | 'security'
  | 'fire'
  | 'accident'
  | 'other';

export type IncidentPriority = 1 | 2 | 3 | 4 | 5;

export interface Incident {
  id: string;
  type: EmergencyType;
  status: IncidentStatus;
  priority: IncidentPriority;
  reporter_id: string;
  reporter_name?: string;
  reporter_email?: string;
  reporter_phone?: string;
  assigned_responder_id?: string;
  assigned_responder_name?: string;
  latitude: number;
  longitude: number;
  location_description?: string;
  campus_block?: string;
  description?: string;
  created_at: string;
  updated_at: string;
  assigned_at?: string;
  responded_at?: string;
  arrived_at?: string;
  resolved_at?: string;
}

export interface IncidentTimelineEvent {
  id: string;
  incident_id: string;
  status: IncidentStatus;
  timestamp: string;
  description?: string;
  actor_id?: string;
  actor_name?: string;
}

export interface IncidentFilter {
  status?: IncidentStatus[];
  type?: EmergencyType[];
  priority?: IncidentPriority[];
  date_from?: string;
  date_to?: string;
  campus_block?: string;
  search?: string;
}

export const INCIDENT_STATUS_LABELS: Record<IncidentStatus, string> = {
  created: 'Created',
  received: 'Received',
  assigned: 'Assigned',
  responding: 'Responding',
  arrived: 'Arrived',
  resolved: 'Resolved',
  cancelled: 'Cancelled',
  escalated: 'Escalated',
  unassigned: 'Unassigned',
  failed: 'Failed',
};

export const EMERGENCY_TYPE_LABELS: Record<EmergencyType, string> = {
  medical: 'Medical',
  security: 'Security',
  fire: 'Fire',
  accident: 'Accident',
  other: 'Other',
};

export const INCIDENT_STATUS_COLORS: Record<IncidentStatus, string> = {
  created: 'bg-gray-100 text-gray-800',
  received: 'bg-blue-100 text-blue-800',
  assigned: 'bg-purple-100 text-purple-800',
  responding: 'bg-amber-100 text-amber-800',
  arrived: 'bg-teal-100 text-teal-800',
  resolved: 'bg-green-100 text-green-800',
  cancelled: 'bg-gray-100 text-gray-500',
  escalated: 'bg-red-100 text-red-800',
  unassigned: 'bg-orange-100 text-orange-800',
  failed: 'bg-red-100 text-red-800',
};

export const EMERGENCY_TYPE_COLORS: Record<EmergencyType, string> = {
  medical: 'bg-red-100 text-red-800',
  security: 'bg-blue-100 text-blue-800',
  fire: 'bg-orange-100 text-orange-800',
  accident: 'bg-yellow-100 text-yellow-800',
  other: 'bg-gray-100 text-gray-800',
};

export type CommunityResponseType =
  | 'offering_assistance'
  | 'escorting_to_safety'
  | 'first_aid_provided'
  | 'eyewitness_report'
  | 'other_assistance';

export interface IncidentCommunityResponse {
  id: string;
  incident_id: string;
  responder_id?: string;
  responder_name?: string;
  response_type: CommunityResponseType;
  message: string;
  created_at: string;
}

export const COMMUNITY_RESPONSE_LABELS: Record<CommunityResponseType, { label: string; icon: string; color: string }> = {
  offering_assistance: { label: 'Assistance Offered', icon: 'handshake', color: 'text-blue-600 bg-blue-50 border-blue-200' },
  escorting_to_safety: { label: 'Escorting to Safety', icon: 'transfer_within_a_station', color: 'text-teal-600 bg-teal-50 border-teal-200' },
  first_aid_provided: { label: 'First Aid Applied', icon: 'medical_services', color: 'text-emerald-600 bg-emerald-50 border-emerald-200' },
  eyewitness_report: { label: 'Eyewitness Report', icon: 'visibility', color: 'text-amber-700 bg-amber-50 border-amber-200' },
  other_assistance: { label: 'Community Assistance', icon: 'volunteer_activism', color: 'text-indigo-600 bg-indigo-50 border-indigo-200' },
};

File: apps/mobile/lib/shared/models/incident.dart
import 'package:equatable/equatable.dart';

class Incident extends Equatable {
  final String id;
  final EmergencyType type;
  final IncidentStatus status;
  final int priority;
  final String? reporterId;
  final String? assignedResponderId;
  final double? latitude;
  final double? longitude;
  final String? locationDescription;
  final String? campusBlock;
  final String? description;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime? assignedAt;
  final DateTime? respondedAt;
  final DateTime? arrivedAt;
  final DateTime? resolvedAt;

  const Incident({
    required this.id,
    required this.type,
    required this.status,
    required this.priority,
    this.reporterId,
    this.assignedResponderId,
    this.latitude,
    this.longitude,
    this.locationDescription,
    this.campusBlock,
    this.description,
    required this.createdAt,
    required this.updatedAt,
    this.assignedAt,
    this.respondedAt,
    this.arrivedAt,
    this.resolvedAt,
  });

  factory Incident.fromJson(Map<String, dynamic> json) {
    return Incident(
      id: json['id'] as String,
      type: EmergencyType.fromString(json['type'] as String),
      status: IncidentStatus.fromString(json['status'] as String),
      priority: json['priority'] as int,
      reporterId: json['reporter_id'] as String?,
      assignedResponderId: json['assigned_responder_id'] as String?,
      latitude: (json['latitude'] as num?)?.toDouble(),
      longitude: (json['longitude'] as num?)?.toDouble(),
      locationDescription: json['location_description'] as String?,
      campusBlock: json['campus_block'] as String?,
      description: json['description'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      assignedAt: json['assigned_at'] != null
          ? DateTime.parse(json['assigned_at'] as String)
          : null,
      respondedAt: json['responded_at'] != null
          ? DateTime.parse(json['responded_at'] as String)
          : null,
      arrivedAt: json['arrived_at'] != null
          ? DateTime.parse(json['arrived_at'] as String)
          : null,
      resolvedAt: json['resolved_at'] != null
          ? DateTime.parse(json['resolved_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'type': type.value,
      'status': status.value,
      'priority': priority,
      'reporter_id': reporterId,
      'assigned_responder_id': assignedResponderId,
      'latitude': latitude,
      'longitude': longitude,
      'location_description': locationDescription,
      'campus_block': campusBlock,
      'description': description,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'assigned_at': assignedAt?.toIso8601String(),
      'responded_at': respondedAt?.toIso8601String(),
      'arrived_at': arrivedAt?.toIso8601String(),
      'resolved_at': resolvedAt?.toIso8601String(),
    };
  }

  bool get isActive => status != IncidentStatus.resolved &&
      status != IncidentStatus.cancelled;

  bool get isAssigned => assignedResponderId != null;

  @override
  List<Object?> get props => [
        id,
        type,
        status,
        priority,
        reporterId,
        assignedResponderId,
        latitude,
        longitude,
        locationDescription,
        campusBlock,
        description,
        createdAt,
        updatedAt,
        assignedAt,
        respondedAt,
        arrivedAt,
        resolvedAt,
      ];
}

enum EmergencyType {
  medical('medical'),
  security('security'),
  fire('fire'),
  accident('accident'),
  other('other');

  const EmergencyType(this.value);
  final String value;

  factory EmergencyType.fromString(String value) {
    return EmergencyType.values.firstWhere(
      (type) => type.value == value,
      orElse: () => EmergencyType.other,
    );
  }

  String get displayName {
    switch (this) {
      case EmergencyType.medical:
        return 'Medical';
      case EmergencyType.security:
        return 'Security';
      case EmergencyType.fire:
        return 'Fire';
      case EmergencyType.accident:
        return 'Accident';
      case EmergencyType.other:
        return 'Other';
    }
  }
}

enum IncidentStatus {
  created('created'),
  received('received'),
  assigned('assigned'),
  responding('responding'),
  arrived('arrived'),
  resolved('resolved'),
  cancelled('cancelled'),
  failed('failed');

  const IncidentStatus(this.value);
  final String value;

  factory IncidentStatus.fromString(String value) {
    return IncidentStatus.values.firstWhere(
      (status) => status.value == value,
      orElse: () => IncidentStatus.created,
    );
  }

  String get displayName {
    switch (this) {
      case IncidentStatus.created:
        return 'Created';
      case IncidentStatus.received:
        return 'Received';
      case IncidentStatus.assigned:
        return 'Assigned';
      case IncidentStatus.responding:
        return 'Responding';
      case IncidentStatus.arrived:
        return 'Arrived';
      case IncidentStatus.resolved:
        return 'Resolved';
      case IncidentStatus.cancelled:
        return 'Cancelled';
      case IncidentStatus.failed:
        return 'Failed';
    }
  }
}
